### PR TITLE
Site: Add possibility for blocks to register a loader function that allows loading additional data

### PR DESCRIPTION
**Pull request by nsams**
This is triggered by getStaticProps by recursively analzying the block data (using block meta data)

This provides another alternatives for blocks to load data. Other alternatives:

- client side fetch (with obvious drawbacks)
- add more data in transformToPlain (needs api change for frontend, difficult to inject dependencies there, also called in admin)
- use react server side components and fetch during runtime (requires next 13 and app router (which we want to switch to anyway), does un-solveable waterfall loading, un-solveable(?) for block preview)

TODO:
- [x] move reusable parts into @comet/cms-site
- [x] avoid waterfall loading (call all loders in pralell)
- [x] block preview is not yet working
- [x] page preview is not yet working
- [x] avoid excessive loader calls while editing and preview updates: #1681 or #1682

### CODE REVIEW

**Review by johnnyomair — COMMENTED** [review]
@nsams first we should merge #1338, right? Also, can this PR target `main`? This should be a non-breaking addition

**Comment by nsams** [comment]
> @nsams first we should merge https://github.com/vivid-planet/comet/pull/1338, right?

yes, this was split out as individual PR (that's why it's not dependent) - once that is merged I will continue working here

> Also, can this PR target main? This should be a non-breaking addition

I hope so

File: packages/site/cms-site/package.json
Comment (nsams) [reply packages/site/cms-site/package.json:58]: I would prefer not bringing in this dependency but don't know an alternative (except typing it as any which also sucks)

File: .changeset/selfish-dolls-beg.md
Comment (johnnyomair) [inline .changeset/selfish-dolls-beg.md:12]:
```suggestion
Add new technique for blocks to load additional data at page level when using SSR

This works both server-side (SSR, SSG) and client-side (block preview).

New functions:

- `registerBlock`: allows registering a loader for a named block
- `recursivelyLoadBlockData`: can be used to call those loaders for a block data tree
```

File: demo/site/src/news/blocks/NewsDetailBlock.tsx
Comment (johnnyomair) [inline demo/site/src/news/blocks/NewsDetailBlock.tsx:33]: Shouldn't you use a variable for this?

File: demo/site/src/pages/preview/admin/page.tsx
Comment (johnnyomair) [inline demo/site/src/pages/preview/admin/page.tsx:16]: Why do you need a ref for this? You could also declare the client outside of the component. Also, should the options of the client be adapted according the current settings in the admin?

File: demo/site/src/recursivelyLoadBlockData.ts
Comment (johnnyomair) [inline demo/site/src/recursivelyLoadBlockData.ts:4]: This includes the complete block meta bundle in the client-side bundle (because of it's usage in `demo/site/src/pages/preview/admin/page.tsx`. Could this be a problem?

File: packages/site/cms-site/src/blockRegistry/blockRegistry.ts
Comment (johnnyomair) [inline packages/site/cms-site/src/blockRegistry/blockRegistry.ts:80]: The client is only passed around here. How about using a generic for this function instead? This would remove the dependency on grapqhl-request in `@comet/cms-site` and would enable using other GraphQL clients (Apollo, urql)?

Comment (nsams) [reply packages/site/cms-site/src/blockRegistry/blockRegistry.ts:80]: Then you also have a a generic in the block loader function that you always have to specify. Maybe we could use module augmentation?

File: demo/site/src/recursivelyLoadBlockData.ts
Comment (nsams) [reply demo/site/src/recursivelyLoadBlockData.ts:4]: True, I haven't thought about that.

I could lazy load it?

File: packages/site/cms-site/src/blockRegistry/blockRegistry.ts
Comment (johnnyomair) [reply packages/site/cms-site/src/blockRegistry/blockRegistry.ts:80]: > Maybe we could use module augmentation?

Thought about this too. Maybe even better!

File: demo/site/src/recursivelyLoadBlockData.ts
Comment (johnnyomair) [reply demo/site/src/recursivelyLoadBlockData.ts:4]: > I could lazy load it?

I just tested it locally in one of our projects. The bundle size shared across all pages is increased about 11kb. I believe it's a good idea to lazy load it.

**Comment by johnnyomair** [comment]
@chernylu FYI 🙂

File: demo/site/src/news/blocks/NewsDetailBlock.tsx
Comment (nsams) [reply demo/site/src/news/blocks/NewsDetailBlock.tsx:33]: yes, I don't know what was wrong with me :)

**Review by johnnyomair — COMMENTED** [review]
What about the dependency on graphql-request? Do you intend to keep it, or try using module augmentation?

**Comment by nsams** [comment]
> What about the dependency on graphql-request? Do you intend to keep it, or try using module augmentation?

first we should decide on that:
#1681 / #1682

**Comment by johnnyomair** [comment]
> > What about the dependency on graphql-request? Do you intend to keep it, or try using module augmentation?
> 
> first we should decide on that: #1681 / #1682

See https://github.com/vivid-planet/comet/pull/1681#pullrequestreview-1900684210

File: demo/api/src/news/blocks/news-detail.block.ts
Comment (johnnyomair) [inline demo/api/src/news/blocks/news-detail.block.ts:4]: We already have [NewsLinkBlock](https://github.com/vivid-planet/comet/blob/next/demo/api/src/news/blocks/news-link.block.ts), should we use that for this use case? IMO it's a more realistic use case.

File: demo/site/src/documentTypes/Page.tsx
Comment (johnnyomair) [inline demo/site/src/documentTypes/Page.tsx:66]: I recently learned that [mutating props is an anti-pattern](https://react.dev/reference/rules/components-and-hooks-must-be-pure#props), so I think we shouldn't do it in Demo, even if it probably will never be a problem here.

File: packages/site/cms-site/src/iframebridge/useBlockPreviewFetch.tsx
Comment (johnnyomair) [inline packages/site/cms-site/src/iframebridge/useBlockPreviewFetch.tsx:12]: Since it isn't a client anymore (which you vehemently pointed out), we should call this `fetchRef` 😁

File: packages/site/cms-site/src/graphQLFetch/fetchInMemoryCache.ts
Comment (johnnyomair) [inline packages/site/cms-site/src/graphQLFetch/fetchInMemoryCache.ts:5]: Should this be part of the public API? Could this be of use in application code somewhere?

File: packages/site/cms-site/src/graphQLFetch/fetchInMemoryCache.ts
Comment (johnnyomair) [inline packages/site/cms-site/src/graphQLFetch/fetchInMemoryCache.ts:14]: This could fail if the body isn't JSON (e.g., form data?). Should we add a try/catch?

File: packages/site/cms-site/package.json
Comment (johnnyomair) [inline packages/site/cms-site/package.json:43]: Why is this needed?

File: demo/site/src/app/block-preview/page/page.tsx
Comment (johnnyomair) [inline demo/site/src/app/block-preview/page/page.tsx:12]: @raphaelblum what do you think about the distinction between fetch (REST) and graphQLFetch (GraphQL)? Is this clear enough?

**Comment by raphaelblum** [comment]
Now that we are using Next 14, couldn't we also have the blocks themselves retrieve data with [streaming + suspense](https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming#streaming-with-suspense)?

File: demo/api/src/news/blocks/news-detail.block.ts
Comment (nsams) [reply demo/api/src/news/blocks/news-detail.block.ts:4]: That NewsDetailBlock is perfect use case, for a NewsLink you'll never use such a loader

File: packages/site/cms-site/src/graphQLFetch/fetchInMemoryCache.ts
Comment (nsams) [reply packages/site/cms-site/src/graphQLFetch/fetchInMemoryCache.ts:5]: Maybe.

As a start I wanted to keep the public API to a minimum.

File: packages/site/cms-site/src/graphQLFetch/fetchInMemoryCache.ts
Comment (nsams) [reply packages/site/cms-site/src/graphQLFetch/fetchInMemoryCache.ts]: yes, absolutely

File: packages/site/cms-site/package.json
Comment (nsams) [reply packages/site/cms-site/package.json]: it is not needed

File: demo/site/src/documentTypes/Page.tsx
Comment (nsams) [reply demo/site/src/documentTypes/Page.tsx]: This variable doesn't contain props, it's named confusing. I renamed it to data.

File: demo/site/src/app/block-preview/page/page.tsx
Comment (raphaelblum) [reply demo/site/src/app/block-preview/page/page.tsx:12]: for me it is, yes.

File: demo/api/src/news/blocks/news-detail.block.ts
Comment (nsams) [reply demo/api/src/news/blocks/news-detail.block.ts:4]: Yes

Comment (johnnyomair) [reply demo/api/src/news/blocks/news-detail.block.ts:4]: If we want to use transformToPlain for use cases like this we need to simplify its usage. Currently we need a lot of complicated stuff in application code to achieve that.

**Review by johnnyomair — APPROVED** [review]

**Review by raphaelblum — APPROVED** [review]

**Pull request merged**
